### PR TITLE
Plotter: update default colors to valid colors.

### PR DESCRIPTION
**Pull request by mtmccrea**
## Purpose and Motivation

Revealed in fa67b746fa6cf9cd29eb42d4fba3f63133e3bf8d, `Plotter` never really handled default colors correctly. When accessing `QtGUI.palettle` colors, it would return unset colors, which would "gracefully" fail by defaulting to black. It seems that correcting the behavior of `-asHSV` made these calls break...ungracefully.

The problem is seen by running 
```supercollider
7.collect({ 15.collect({ rrand(0,1.0) }) }).plot.superpose_(true)
```
The proposed defaults access `QtGUI.palette.highlight`, which returns a valid color, whereas `QtGUI.palette.windowText`, used before, didn't.

I've also reduced the default `plotColor` down to one color. Four colors was a rather random number of colors to default to, and in any case the user can set them explicitly in `Plot` (more easily through `Plotter` in a pending PR merge — #4459).

## Types of changes

- Bug fix

## To-do list

- [x] Code is tested
- [x] All tests are passing
~~- [ ] Updated documentation~~
- [x] This PR is ready for review

### CODE REVIEW

**Comment by mtmccrea** [comment]
Hmmm.. Looks like the a test isn't passing:
https://travis-ci.org/supercollider/supercollider/jobs/564793041#L4506

>  [!] TestPlotter:test_suppliedBounds (0.0s) (Did not complete.)
           ! Error: "ERROR: Message 'use' not understood."

But I don't see `test_suppliedBounds` in `TestPlotter.sc`. Where can I find this test to see why it's failing?

**Comment by brianlheim** [comment]
> Where can I find this test to see why it's failing?

These tests are pull from CommonTests quark (unfortunately). You may need to make a PR against that repo to fix it.

**Comment by mtmccrea** [comment]
Still troubleshooting what's breaking in the test... I'll post back here when it's ready for review.

**Comment by dyfer** [comment]
IIUC we are in the process of changing (fixing?) tests on CI so they use the built-in testsuite, instead of the CommonTests quark: #3132. After that's merged, this PR would probably need to be rebased to include that change (?) and trigger a new Travis build.

**Comment by mtmccrea** [comment]
Hi @dyfer and @brianlheim, finally finding some time to come back around to SC work during the holiday break...
So I think this may be breaking on account of a previous change to `Plot`:
dea216a1477ecbdd75efcfb5d7d2e81dcbd037c4
The error in the travis build is on the `use` method, which is being called on `GUI.skin.at(\plot)`

https://github.com/supercollider/supercollider/blob/3037bf299bf21618bdf11fce53bf53427e84f13a/SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc#L39-L44

However, skin isn't initialized if `Platform.hasQt.not`

https://github.com/supercollider/supercollider/blob/3037bf299bf21618bdf11fce53bf53427e84f13a/SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc#L13-L16

So I expect this is the case when testing non-Qt travis build configurations?
It looks like the change made in dea216a1477ecbdd75efcfb5d7d2e81dcbd037c4 may have been hasty without consideration of a user (or test) trying to use `Plotter` without Qt. 

Not sure what the appropriate solution is. I would think that if testing a non-Qt build, tests using Qt should be omitted.

File: SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc
Comment (telephon) [inline SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: if it's not needed, perhaps better remove and explain why it can be removed?

Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: @telephon i was just experimenting with whether this avoided the break, I wasn't planning to keep this change. in any case it didn't appear to help, so i'm unsure what exactly is going wrong

Comment (telephon) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: ah good!

File: SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc
Comment (brianlheim) [inline SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: @mtmccrea -- are these methods from an extension or something?  they are not in vanilla SC, and that appears to be why the unit tests are failing. i get this message on startup:

```
ERROR: Message 'sat_' not understood.
RECEIVER:
Instance of Color {    (0x115a482b8, gc=6C, fmt=00, flg=00, set=02)
  instance variables [4]
    red : Float 0.298039   13131313 3FD31313
    green : Float 0.498039   DFDFDFE0 3FDFDFDF
    blue : Float 0.749020   F7F7F7F8 3FE7F7F7
    alpha : Float 1.000000   00000000 3FF00000
}
ARGS:
   Float 0.200000   9999999A 3FC99999
CALL STACK:
	DoesNotUnderstandError:reportError
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand
		arg this = <instance of Color>
		arg selector = 'sat_'
		arg args = [*1]
	< closed FunctionDef >  (no arguments or variables)
	Function:doOnStartUp
		arg this = <instance of Function>
	ArrayedCollection:do
		arg this = [*6]
		arg function = <instance of Function>
		var i = 4
	List:do
		arg this = <instance of List>
		arg function = <instance of Function>
	Meta_StartUp:run
		arg this = <instance of Meta_StartUp>
	Main:startup
		arg this = <instance of Main>
		var didWarnOverwrite = false
^^ The preceding error dump is for ERROR: Message 'sat_' not understood.
RECEIVER: Color(0.29803921568627, 0.49803921568627, 0.74901960784314)
```

Comment (dyfer) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: `.sat_` seems to come from the `wslib` quark...

Comment (brianlheim) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: got it. good lesson for everyone, don't do development work with extensions enabled :)

Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: ahhh! thanks for catching this @brianlheim (and  @dyfer)!
strange that this wasn't the error thrown in the test. I'll update the functionality with something built-in.

Comment (brianlheim) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: well, this code isn't exercised in the test, that's the problem

**Comment by brianlheim** [comment]
thanks @mtmccrea! i'm going to mark all the comments about CI failure as off topic so this PR is easier to read. not using QtGUI.palette for this kind of concerns me -- isn't there any other palette element we could use to populate these values?

File: SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc
Comment (mtmccrea) [inline SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: @brianlheim QtGUI.palette is still used here, just refactored into a variable.

**Comment by mtmccrea** [comment]
@brianlheim `QtGUI.palette` is still used, perhaps I'm not understanding your concern. Could you elaborate? I'm not that familiar with palettes, but happy to make changes if you prefer.

>  i'm going to mark all the comments about CI failure as off topic so this PR is easier to read.
Sounds good. Those failures are still a mystery to me, hopefully they're not indicative of other underlying problems in the CI tests.

File: SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc
Comment (brianlheim) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: fontColor is hard coded though

Comment (jamshark70) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: > fontColor is hard coded though

Not a trivial point. On my system, the base color is dark, so gray 0.3 is likely to be dark text on a dark background.

Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: Ah, I see the issue, thanks for spotting it. I'll look into palettes to see if I can come up with something that works across palettes.

**Comment by dyfer** [comment]
@mtmccrea any progress on this?

**Comment by joshpar** [comment]
@mtmccrea and @dyfer - checking in again to see if this is ready. I think we will try to get it in the 3.11 release if we can resolve anything left over this week.

**Comment by mtmccrea** [comment]
Hi @joshpar yes I have a fix now but I was intending to do a bit more testing. It took some time because it has uncovered other issues with plotter, and I needed to consider if they fit in the scope of this PR. If there’s a time crunch I could push what I have for review and file tangential issues for later.
What’s the timeline for 3.11?

**Comment by joshpar** [comment]
I think beta by this weekend. This, however, isn’t a feature change (in my opinion), and could be wrapped up by RC1 (that I’m guessing will be about a month down the road).
Thanks!

**Comment by mtmccrea** [comment]
While I haven't had time to wrap up testing on this, I thought I'd at least see what folks think of the proposed new default colors. 

@brianlheim and @jamshark70 brought up the good point that using palettes would be sensible. So after much futzing, here's what the default colors would look like with the updated version, for  `system`, `dark`, and `light`, palettes:

Screen Shot 2020-01-31 at 12 18 49 AM

Thoughts?

**Comment by brianlheim** [comment]
that looks good to me!

**Comment by dyfer** [comment]
@mtmccrea I also like how it looks.
If the `system` palette follows OS settings (I assume it would?), I'd just make sure that it looks fine when using the light color scheme in the OS.

**Comment by mtmccrea** [comment]
>If the system palette follows OS settings

Doesn't seem to... I get the `light` palette regardless of my system settings.

**Comment by dyfer** [comment]
> Doesn't seem to... I get the `light` palette regardless of my system settings.
> 

Sorry I wasn't clear. When the `system` palette is used, do the actual colors change with the change of the system theme between light and dark? And with or without restarting SC?

**Comment by brianlheim** [comment]
@mtmccrea @dyfer where are we with this? it looks like @mtmccrea just needs to push the changes from this comment? https://github.com/supercollider/supercollider/pull/4511#issuecomment-593274864

**Comment by mtmccrea** [comment]
I thought I'd have a quick look at this again to see if I could wrap it up. It's been a while so I'm going off of notes I made when my head was in the problem:

I realized that the original author of `Plotter` intended that plot attributes can be set from an Event, `GUI.skin(\plot)`, and then reset by calling `Plot.initClass`. So the most recent changes to the PR may look odd, but it's to preserve the original design:

If you change the global QT palette and want it to apply it to new plots, you need to call `Plot.initClass` after setting the palette and before creating the plot. The logic appears to be that
you can set the `GUI.skin(\plot)` properties directly, and you only want to reset them explicitly, hence `Plot.initClass`. Seems an odd paradigm to me.

Regardless of that being a good idea or not, `Plot` isn't a class intended for direct user interaction anyway, `Plotter` is. So this functionality, if kept, should probably eventually move to that class.

In any case a quick test to try on your system would be:
```
// default out the gate:
p = Array.geom(4, 15, 4).collect({|cnt| cnt.collect({rrand(0,1.0)})}).plot

QtGUI.palette = QPalette.system;
Plot.initClass;
p = Array.geom(4, 15, 4).collect({|cnt| cnt.collect({rrand(0,1.0)})}).plot

QtGUI.palette = QPalette.dark;
Plot.initClass;
p = Array.geom(4, 15, 4).collect({|cnt| cnt.collect({rrand(0,1.0)})}).plot

QtGUI.palette = QPalette.light;
Plot.initClass;
p = Array.geom(4, 15, 4).collect({|cnt| cnt.collect({rrand(0,1.0)})}).plot
```
It looks like since the last time i updated the PR, the Qt color definitions have changed already (you can preview default Palette colors with [this script](https://gist.github.com/mtmccrea/2bd28b8ffc64ca0ea8f4f93dbf0076d7)), but the defaults I've made still look OK i think.

It would be a good idea to draw attention to this change if it makes it into the distribution, so that people can try it out with their UIs, especially on other OSs (I only tested on MacOS).

**Comment by telephon** [comment]
As the original author of the thing, please just go ahead and rethink what needs to be rethought! It was just a first version to be modified. Thank you for the improvements...

**Comment by mtmccrea** [comment]
Thanks @telephon I may have a look at the larger architecture in the future, but for now I’ll limit the scope of the PR because I haven’t been able to reliably follow up in the past months. 

I’m happy to make adjustments to the current changes as needed though!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by telephon** [comment]
OK, good!

**Comment by dyfer** [comment]
@mtmccrea @telephon what's the status of this PR?

**Comment by mtmccrea** [comment]
> @mtmccrea @telephon what's the status of this PR?

It's ready for review. Because this is a change to the visual appearance of plots, I would hope others might run it and see if they're ok with the colors. I posted a snippet one might run to see it:
https://github.com/supercollider/supercollider/pull/4511#issuecomment-644398846

**Comment by dyfer** [comment]
I'd like to see if we can get wrap this up.

I tested this and can confirm that the changes introduced work as advertised on macOS, as well as Ubuntu 20.04 and Windows 10, with changes to the Palette being reflected in the plots. 
My notes (for the future):
- I personally wouldn't mind having plotter with multiple colors when superimposed, but let's leave that for another PR.
- BTW It is tricky that one needs to run `Plot.initClass` after changing the `QPalette` (and then redrawing the plots, while some elements of the GUI update instantaneously), but that's also for another PR I think.
- This PR probably should _not_ be cherry picked into 3.11.2 (before anyone suggests it), as it removes the (undocumented anyway) `Plot -gui` method.

@telephon since you were reviewing this, would you mind giving more feedback or possibly approving this PR?

**Review by telephon — CHANGES_REQUESTED** [review]

**Comment by mtmccrea** [comment]
After rebasing, it looks like this will need some more work, so unfortunately it needs to wait.

I made a [draft change](https://github.com/mtmccrea/supercollider/commit/b003f147ad09002f0bd24f11bf01ff116471b8af) that addresses @telephon's suggestion, but I'll need to work with it a bit more and run through the many tests for plotter I have. It's really easy to overlook cases with Plotter.

**Comment by telephon** [comment]
> Btw @telephon, did you mean to suggest that `doOnStartUp` be a class method? It wasn't readily working otherwise, but I may have misunderstood...

Yes, sorry that was unclear – yes, you can call it also on the class.

**Comment by mtmccrea** [comment]
@dyfer 
> Sorry I wasn't clear. When the system palette is used, do the actual colors change with the change of the system theme between light and dark? And with or without restarting SC?

Yes. The `system` palette does indeed change when the OS theme changes. No restart or recompile required.

>IIUC from your previous image, the system palette seems to be close to the dark one.

This is because the OS's "dark" theme (dynamic as the OS evolves) will be similar in theme but not in actual colors to the SC-defined `QPalette.dark` specification (static, hard-coded).

https://github.com/supercollider/supercollider/blob/18c4aad363c49f29e866f884f5ac5bd35969d828/SCClassLibrary/Common/GUI/Base/QPalette.sc#L13-L32

Looking into this brought about a bit more tuning so colors are robust across system and palette changes, as well as spotting an invisible bug in how the `\plot` value in the `skins` dictionary  was being set incorrectly (it was adding it as a property of the current `skin` rather, as an entry into the `GUI.skins` dictionary.)

I think all concerns have been addressed and this PR is ready to go! 
So I'll ping previous reviewers @telephon, @joshpar, and @dyfer!

**Comment by mtmccrea** [comment]
Update: this had failing tests (random configuration errors in CI) which prevented all tests from running. After re-running those jobs, it revealed one legit failing test, and that's been resolved. 
@dyfer

**Review by dyfer — APPROVED** [review]
Thanks @mtmccrea, looks and great!
Thanks @telephon for reviewing

I previously tested on multiple platforms and now re-confirmed that everything works on macOS.

**Comment by dyfer** [comment]
Before merging, I have one final comment: this PR _removes_ the `Plotter -gui` method. This method was not used and only returned `nil`, AFAICT. However, calling this method after this PR will not result in an error: instead the call is propagated to the parent classes and returns an `ObjectGui`. IMO this is fine, but I'll wait for feedback on this before merging. @mtmccrea @telephon @joshpar

File: SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc
Comment (dyfer) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: I believe this was resolved, is that right @telephon ?

**Comment by telephon** [comment]
my 2 cents: would be good to mention in the release notes, otherwise fine!

File: SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc
Comment (mtmccrea) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: bb61d321124622d2ba1f61a863a5c349ddd6fc3f (lines 44-48)

**Comment by mtmccrea** [comment]
>  this PR removes the Plotter -gui method. This method was not operational and only ever returned nil, AFAICT.

Yes, the class never actually set  `gui`, so the getter method was superfluous.

**Comment by dyfer** [comment]
@telephon if your concerns have been addressed, would you mind approving this? (there are "changes requested" by you)

File: SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc
Comment (telephon) [reply SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc]: thanks!

**Review by telephon — APPROVED** [review]

**Pull request merged**
